### DISCUSSION TITLE
Algorithm complexity of neighbor list

### DISCUSSION THREAD

**Discussion by lhycms** [deepmodeling/deepmd-kit#2691] [Q&A] [answered]
I want to  use neighbor list do something. But i can't figure out the implementation in https://github.com/deepmodeling/deepmd-kit/blob/master/source/lib/src/neighbor_list.cc. 
What confuse me is the algorithm complexity of this implementation. Is it $O(N^2)$?

**Comment by njzjz** [comment]
Yes.

**Reply by lhycms** [reply]
Thanks for your help.

**Comment by wanghan-iapcm** [comment]
several implementations are provided, which one do you mean?

**Reply by lhycms** [reply]
I see there are many implementation of `builld_nlist()`. Is there a implementation similar to Lammps, whose algorithm complexity is $O(N)$?

**Reply by njzjz** [reply, marked as answer]
To be clear, although there are several implementations, only `build_nlist_cpu` is actually used, whose complexity is $O(N^2)$.

**Reply by lhycms** [reply]
Got it. Thanks
